metals/src/main/scala/scala/meta/internal/metals/ammonite/Ammonite.scala

```diff
@@ -0,0 +1,435 @@
+  ): Either[AmmoniteFetcherException, (AmmCommand, AbsolutePath)] = {
+    val it = path.toInputFromBuffers(buffers).value.linesIterator
+    val versionsOpt = VersionsOption.fromScript(it)
+    val versions = versionsOpt
+      .orElse(lastImportVersions)
+      .getOrElse(
+        // TODO Warn if any of these is used
+        AmmVersions
+          .default()
+          // temporary, remove once Ammonite is available for 2.12.11
+          // .withScalaVersion(BuildInfo.scala212)
+          .withScalaVersion("2.12.10")
+          // do not merge, remove once lihaoyi/Ammonite#1053 is merged
+          .withAmmoniteVersion("2.0.5-bsp-4")
+      )
+    val res = AmmoniteFetcher(versions)
+      .withInterpOnly(false)
+      .withProgressBars(false)
+      .withResolutionParams(
+        coursierapi.ResolutionParams.create().withScalaVersion("2.12.10")
+      )
+      // do not merge, remove once lihaoyi/Ammonite#1053 is merged
+      .withTransformFetch(Some { fetch =>
+        fetch.addRepositories(
+          coursierapi.MavenRepository
+            .of("https://dl.bintray.com/alexarchambault/metals-amm")
+        )
+      })
+      .command()
+    res match {
+      case Left(e) =>
+        scribe.error(
+          s"Error getting Ammonite ${versions.ammoniteVersion} (scala ${versions.scalaVersion})",
+          e
+        )
+        Left(e)
+      case Right(command) =>
+        lastImportVersions = VersionsOption(
+          Some(versions.ammoniteVersion),
+          Some(versions.scalaVersion)
+        )
+        Right((command, path))
+    }
+  }
+
+  private def isMillBuildSc(path: AbsolutePath): Boolean =
+    path.toNIO.getFileName.toString == "build.sc" &&
+      // for now, this only checks for build.sc, but this could be made more strict in the future
+      // (require ./mill or ./.mill-version)
+      buildTools().isMill
+
+  def maybeImport(path: AbsolutePath): Future[Unit] =
+    if (path.isAmmoniteScript && !isMillBuildSc(path) && !loaded(path)) {
+      val futureResp = languageClient
+        .showMessageRequest(Messages.ImportAmmoniteScript.params())
+        .asScala
+      futureResp.onComplete {
+        case Failure(e) => scribe.warn("Error requesting Ammonite import", e)
+        case Success(resp) =>
+          resp.getTitle match {
+            case Messages.ImportAmmoniteScript.doImport =>
+              start(Some(path)).onComplete {
+                case Failure(e) =>
+                  scribe.warn(s"Error importing Ammonite script $path", e)
+                case Success(_) =>
+              }
+            case _ =>
+          }
+      }
+      futureResp.ignoreValue
+    } else
+      Future.unit
+
+  def start(doc: Option[AbsolutePath] = None): Future[Unit] = {
+
+    disconnectOldBuildServer().onComplete {
+      case Failure(e) =>
+        scribe.warn("Error disconnecting old Ammonite build server", e)
+      case Success(()) =>
+    }
+
+    // TODO Look at buffers.open rather than focusedDocument?
```

Comment:
Focused document would be when we run the command by hand? If so I think it's fine to have it this way.